sleigh/sleigh-compile/src/matcher/mod.rs

```diff
@@ -19,421 +21,304 @@ pub(crate) fn build_sequential_matcher(
     Ok(SequentialMatcher { cases, token_size })
 }
 
-enum Index {
-    Token(u8),
-    Context(u8),
+struct IndexLinkedList<'a, T> {
+    head: Option<usize>,
+    order: Vec<Option<usize>>,
+    elements: &'a [T],
 }
 
-struct BitConstraint {
-    index: Index,
-    ones: u32,
-    zeroes: u32,
-    unconstrained: u32,
+impl<'a, T> IndexLinkedList<'a, T> {
+    fn new_empty(elements: &'a [T]) -> Self {
+        let order = Vec::with_capacity(elements.len());
+        Self { head: None, order, elements }
+    }
+    fn final_order(self) -> impl Iterator<Item = usize> {
```

Comment:
I think this this fails to include very first element.

Something like this:

```rust
impl Iterator for Iter {
    type Item = usize;
    fn next(&mut self) -> Option<Self::Item> {
        let next = self.1?;
        self.1 = self.0[next];
        Some(next)
    }
}
```
Would work I think.

I think this could also be implemented using the [`std::iter::successors`](https://doc.rust-lang.org/stable/std/iter/fn.successors.html) function. i.e:

```
    fn final_order(self) -> impl Iterator<Item = usize> {
        std::iter::successors(self.head, move |i| self.order[*i])
    }
```

Reply:
Fixed